File: app/src/main/java/com/suhane/flickr/ui/base/BaseActivity.java
package com.suhane.flickr.ui.base;

import android.app.ProgressDialog;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.widget.ProgressBar;

import com.suhane.flickr.R;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Optional;
import butterknife.Unbinder;

/**
 * Created by shashanksuhane on 04/02/18.
 */

public abstract class BaseActivity extends AppCompatActivity {

    private Unbinder unbinder;

    @BindView(R.id.progressbar)
    protected ProgressBar progressBar;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(getLayout());
        unbinder = ButterKnife.bind(getActivity());
    }

    @Override
    protected void onDestroy() {
        unbinder.unbind();
        super.onDestroy();
    }

    protected void showProgress() {
        progressBar.setVisibility(View.VISIBLE);
    }

    protected void hideProgress() {
        progressBar.setVisibility(View.INVISIBLE);
    }

    protected abstract void init();
    protected abstract void reset();
    protected abstract int getLayout();
    protected abstract BaseActivity getActivity();
}

File: app/src/main/java/com/suhane/flickr/ui/base/BaseActivity.kt
package com.suhane.flickr.ui.base

import android.os.Bundle
import android.support.annotation.Nullable
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.widget.ProgressBar

import com.suhane.flickr.R

import butterknife.BindView
import butterknife.ButterKnife
import butterknife.Unbinder

/**
 * Created by shashanksuhane on 04/02/18.
 */

abstract class BaseActivity : AppCompatActivity() {

    private var unbinder: Unbinder? = null

    @BindView(R.id.progressbar) @Nullable
    @JvmField
    var progressBar: ProgressBar? = null

    protected abstract val layout: Int
    protected abstract val activity: BaseActivity

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(layout)
        unbinder = ButterKnife.bind(activity)
    }

    override fun onDestroy() {
        unbinder!!.unbind()
        super.onDestroy()
    }

    protected fun showProgress() {
        progressBar!!.visibility = View.VISIBLE
    }

    protected fun hideProgress() {
        progressBar!!.visibility = View.INVISIBLE
    }

    protected abstract fun init()
    protected abstract fun reset()
}
